### PR TITLE
Python SDK Async Client Update

### PR DESCRIPTION
**Pull request by johnstonematt**
#### Doc Strings
Updated doc strings to match the documentation in the [Aptos API Spec](https://aptos.dev/nodes/aptos-api-spec/).

#### New Methods for API Endpoints:
Added new methods for the following endpoints:
* Account:
  * `account_module()`
  * `account_modules()`
* Blocks:
  * `blocks_by_height()`
  * `blocks_by_version()`
* Events:
  * `events_by_creation_number()`
  * `events_by_event_handle()`
* Transactions:
  * `transactions()`
  * `transactions_by_account()`
* View:
  * `view()`

#### Removed Code Duplication
* Added `_get` and `_post` utility methods.
* Removed `simulate_bcs_transaction()` duplication within `simulate_transaction()`.

#### Gitignore
Added a couple things to the Python SDK `.gitignore`, namely `.idea/` (for IntelliJ developers) and `playground.py` (a testing script).

### Test Plan
There is currently no testing for the `AsyncClient` in the Python SDK. I would like to implement tests using [Pytest](https://docs.pytest.org/en/7.4.x/) in a later PR.

### CODE REVIEW

File: ecosystem/python/sdk/.gitignore [hunk 1]
```diff
@@ -161,5 +161,7 @@ cython_debug/
 #  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
-#.idea/
+.idea/
 
+# testing/fidling script:
+playground.py
```
Comment (gregnazario) [inline ecosystem/python/sdk/.gitignore:167]: I would likely revert this change for the purposes of the Pull request and leave it how it is.

File: ecosystem/python/sdk/aptos_sdk/async_client.py [hunk 2]
```diff
@@ -527,6 +784,72 @@ async def transfer_object(
...
+        :param ledger_version: Ledger version to get state of account. If not provided, it will be the latest version.
+        :returns: Execution result.
+        """
+        response = await self._post(
+            endpoint="view",
+            params={
+                "ledger_version": ledger_version,
+            },
+            headers={
+                "Accept": "application/json, application/x-bcs",
```
Comment (gregnazario) [inline ecosystem/python/sdk/aptos_sdk/async_client.py:812]: I would probably just leave this to `applicaiton/json` for now.  Mainnet doesn't have BCS support quite yet, and I'm sure it will probably change abruptly at some point (when both are available, it'll just switch to bcs on you.

File: ecosystem/python/sdk/aptos_sdk/async_client.py [hunk 3]
```diff
@@ -527,6 +784,72 @@ async def transfer_object(
...
+        data: Optional[Dict[str, Any]] = None,
+    ) -> httpx.Response:
+        # format params:
+        params = {} if params is None else params
+        params = {key: val for key, val in params.items() if val is not None}
+        return await self.client.post(
+            url=f"{self.base_url}/{endpoint}",
+            params=params,
+            headers=headers,
+            # data=data,
```
Comment (gregnazario) [inline ecosystem/python/sdk/aptos_sdk/async_client.py:840]: Can remove this line

File: ecosystem/python/sdk/aptos_sdk/async_client.py [hunk 4]
```diff
@@ -259,26 +474,10 @@ async def simulate_transaction(
...
-            }
-
-        response = await self.client.post(
-            f"{self.base_url}/transactions/simulate",
-            params=params,
-            headers=headers,
-            content=signed_transaction.bytes(),
+        return await self.simulate_bcs_transaction(
+            signed_transaction=SignedTransaction(transaction, authenticator),
+            estimate_gas_usage=estimate_gas_usage,
```
Comment (gregnazario) [inline ecosystem/python/sdk/aptos_sdk/async_client.py:479]: I don't think `estimate_gas_usage` is used in `submit_bcs_transaction`?  Might have to port this over from the old code here.

**Review by gregnazario — COMMENTED** [review]
Looks strictly cleaner, and much better docs than before.

Just a few cleanup comments, and the gas estimation looks like it disappears?

File: ecosystem/python/sdk/aptos_sdk/async_client.py (hunk 4 above)
Comment (johnstonematt) [reply ecosystem/python/sdk/aptos_sdk/async_client.py:479]: No it's not. `estimate_gas_usage` is only used in `simulate_transaction` and `simulate_bcs_transaction`, not `submit_transaction` and `submit_bcs_transaction`. Is this an issue?

File: ecosystem/python/sdk/.gitignore (hunk 1 above)
Comment (johnstonematt) [reply ecosystem/python/sdk/.gitignore:167]: I can remove the `playground.py` and simply delete the file, but IntelliJ always inserts a `.idea` folder which contains IntelliJ project specific settings, unique to each user, so I think this should be included in the .gitignore.

Can read more [here](https://rider-support.jetbrains.com/hc/en-us/articles/207097529-What-is-the-idea-folder-#:~:text=Follow&text=2023%2006%3A46-,The%20.,history%20and%20currently%20selected%20configuration.).

Comment (gregnazario) [reply ecosystem/python/sdk/.gitignore:167]: Yeah, fair.  I don't know who added the original comment around .idea.  Just let's remove playground.

File: ecosystem/python/sdk/aptos_sdk/async_client.py (hunk 4 above)
Comment (gregnazario) [reply ecosystem/python/sdk/aptos_sdk/async_client.py:479]: I read wrong, read it as submit instead of simulate.  My mistake

**Review by gregnazario — APPROVED** [review]

**Review by davidiw — COMMENTED** [review]
Overall looks good. We can add to the readme how to test:
* make test -- runs basic unit tests
* make fmt -- autofix some lints and other formatting quirks
* make lint -- identifies non-auto fix lints
* make examples_cli -- uses a local aptos cli binary to spin up a local testnet and run all the examples -- this tests much of the sdks api calls

Could you run through those? These tests seem to drop off frequently. The examples hasn't been added yet, but really ought to asap...

**Comment by davidiw** [comment]
@johnstonematt , lemme know when I should review again. Thanks for your efforts!

**Comment by johnstonematt** [comment]
@davidiw I ran all the commands you mentioned, all of them passed except `make lint` which failed with 1 missing-import error in `aptos_sdk/secp256k1_ecdsa.py`. 

I ran the test commands in the `main` branch as well (un-edited by my PR) and saw the same results - all passed except the missing-import error in the same file.

Let me know if you would like me to copy and paste the results of the commands.

File: ecosystem/python/sdk/aptos_sdk/async_client.py [hunk 5]
```diff
@@ -527,6 +782,73 @@ async def transfer_object(
...
+            raise ApiError(response.text, response.status_code)
+
+        return response.content
+
+    async def _post(
+        self,
+        endpoint: str,
+        params: Optional[Dict[str, Any]] = None,
+        headers: Optional[Dict[str, Any]] = None,
+        data: Optional[Dict[str, Any]] = None,
```
Comment (davidiw) [inline ecosystem/python/sdk/aptos_sdk/async_client.py:829]: wouldn't it be easier to just make this not optional and instead return a dict() by default instead?

beyond that, everything seems good

Comment (johnstonematt) [reply ecosystem/python/sdk/aptos_sdk/async_client.py:829]: With regard to making them not optional, I don't think so, as not every call would necessarily need to specify those (similar to the `.post` method in the underlying `httpx.AsyncClient`).

With regard to returning a `dict`, the function instead returns the actual `Response` object from the `httpx` library, such that any parent function using the method can understand the context around the response (i.e., the error code in the case of errors). 

Ideally, I would think that in future error handling and general "response processing" would be automatically handled in these utility methods (`_get()`, `_post()`, `_delete()`), but for now the method can just return the full response so that processing can be done in whatever parent method is calling the utility function.

Does that answer your question?

Comment (davidiw) [reply ecosystem/python/sdk/aptos_sdk/async_client.py:829]: I think the word optional is overloaded here...

more explicitly
```suggestion
        params: Dict[str, Any] = dict(),
```

as this eliminates line 832

similarly below on line 845

Comment (johnstonematt) [reply ecosystem/python/sdk/aptos_sdk/async_client.py:829]: That is a mutable default argument, which is said to be bad practice as it can cause very sneaky bugs

Comment (davidiw) [reply ecosystem/python/sdk/aptos_sdk/async_client.py:829]: Hmm... I thought `dict()` and `list()` wouldn't have the same issues, I stand corrected: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python

**Review by davidiw — APPROVED** [review]

File: ecosystem/python/sdk/aptos_sdk/async_client.py (hunk 5 above)
Comment (johnstonematt) [reply ecosystem/python/sdk/aptos_sdk/async_client.py:829]: For a second I thought the same, but I imagine `dict()` would run and execute to `{}` before the function is actually defined (as opposed to only running when the function is called), so the effect would be the same as putting `{}` as the default argument.
